app/ts/components/HashRouter.tsx

```diff
@@ -0,0 +1,65 @@
+import { ComponentChild, FunctionalComponent, toChildArray, VNode } from 'preact'
+import { useEffect, useState } from 'preact/hooks'
+
+export type RouteProps = {
+	path: string
+	component: FunctionalComponent
+}
+
+export type RouterProps = {
+	children: VNode<RouteProps> | VNode<RouteProps>[]
+	onHashChange?: (event: Event) => void
+}
+
+function stripLeadingSlash(urlString: string) {
+	if (urlString.charAt(0) !== '/') return urlString
+	return urlString.substring(1)
+}
+
+function matchPattern(pattern: string, str: string): boolean {
+	const patternParts = pattern.split('/')
+	const strParts = str.split('/')
+
+	if (patternParts.length !== strParts.length) return false
+
+	for (let i = 0; i < patternParts.length; i++) {
+		const patternPart = patternParts[i]
+		const strPart = strParts[i]
+		if (patternPart.startsWith(':')) continue
+		if (patternPart !== strPart) return false
+	}
+
+	return true
+}
+
+function useOnHashChange(handler?: (event: Event) => void) {
+	useEffect(() => {
+		if (!handler || typeof handler !== 'function') return
+		window.addEventListener('hashchange', handler)
+		return () => window.removeEventListener('hashchange', handler)
+	}, [])
+}
+
+export const Route = ({ component: Component }: RouteProps) => {
+	return <Component />
+}
+
+export const Router = ({ children, onHashChange }: RouterProps) => {
+	const [activeRoute, setActiveRoute] = useState(getActiveRoute())
+
+	function getActiveRoute() {
+		// ts does not assert correctly if these are combined into one function
```

Comment:
I might be missing something on typing array filtering, tho this works
```ts
const isVNode = (child: ComponentChild): child is VNode<any> => child instanceof Object
const isRouteComponent = (child: VNode<any>): child is VNode<RouteProps> => 'path' in child.props && 'component' in child.props
const childArray = toChildArray(children).filter(isVNode).filter(isRouteComponent)
// childArray is type VNode<RouteProps>

```
tho i expected this would also work
```ts
const isRouteComponent = (child: ComponentChild): child is VNode<RouterProps> => child instanceof Object && 'path' in child.props && 'component' in child.props
const childArray = toChildArray(children).filter(isRouteComponent)
// childArray is 'string' | 'number' | VNode<{}>
// which is from toChildArray

```

Reply:
The root of the problem is:
```ts

Reply:
opted in with custom typing `children` component as object array shying away from `toChildArray`